Add auto-generated version.h header to API headers.

est77 opened the pull request:
A version.h header is now auto-generated with the appropiate defines when building.
This is useful for clients of the C++ API to keep their code working with different versions of appleseed.
I removed the code to get the version numbers from github, so the problem when building from a tarball 
is also fixed. Version numbers are now explicit in the CMakeLists.txt file.
I'll update the documentation related to releasing appleseed when this PR is merged.

dictoon commented on CMakeLists.txt:
```diff
@@ -122,30 +122,16 @@ endif ()
+set (appleseed_version_patch 0)
+set (appleseed_version_alpha 24)
 
+configure_file (${PROJECT_SOURCE_DIR}/src/appleseed/renderer/api/version.h.in
```
I would suggest to place versioning files into `foundation/core/` instead.

est77 replied:
I need the header in renderer/api. I could generate the version file in foundation/core and create a renderer/api/version.h header that includes it, but I think it's better to have the version in only one place.

dictoon commented on src/appleseed/foundation/core/appleseed.cpp:
```diff
@@ -30,6 +30,9 @@
 // Interface header.
 #include "appleseed.h"
 
+// appleseed.renderer headers.
```
Including appleseed.renderer headers from appleseed.foundation is not allowed. See my remark above about the location of the versioning files.

est77 replied:
What about moving this to renderer/core for example. It's not really related to foundation.

dictoon replied:
Sounds good.

dictoon commented on CMakeLists.txt:
```diff
@@ -122,30 +122,16 @@ endif ()
+set (appleseed_version_major 1)
+set (appleseed_version_minor 1)
+set (appleseed_version_patch 0)
+set (appleseed_version_alpha 24)
```
I would suggest replacing `appleseed_version_alpha` by `appleseed_version_maturity` set to "alpha-24".

est77 replied:
I need a number that I can compare. A string wont work.

dictoon replied:
What about this then:

`appleseed_version_maturity_level`:
* 0 = alpha
* 1 = beta
* 2 = release candidate
* 3 = stable

`appleseed_version_candidate` = 24

est77 replied:
My motivation for adding this is to be able to check the version numbers in code that uses the apleseed library and decide from that which features are available. Keeping client code backwards compatible but being able to use new features when they are added.
Having five version numbers is just too much. We need to simplify a bit.

est77 replied:
What if we make the next stable release 1.2.0 and remove the candidate / alpha number?

dictoon replied:
Why not, but that would mean that the next release is the first stable release. Are we ready for that?

est77 replied:
I meant, next one would be 1.1.1, 1.1.2, ... and 1.2.0 when we are ready.

dictoon replied:
In the absence of an "alpha" or "beta" tag, as soon as 1.x.x is reached the software is considered stable, according to the semantic versioning conventions.

This is why I stand by the proposal I outlined in our private conversation: We stick to 1.1.0 for now, with the addition of a candidate number (currently at 25). When we are ready to release our first stable version, we give it the version number 1.2.0 and we stop talking about candidate numbers (but we can and should keep them internally). From that moment on, we stick to the semantic versioning rules. So the release after 1.2.0 will be either 1.2.1, 1.3.0 or 2.0.0 depending on what it breaks.

dictoon commented on src/appleseed/renderer/api/version.h.in:
```diff
@@ -0,0 +1,50 @@
+    APPLESEED_VERSION_MINOR * 100 + \
+    APPLESEED_VERSION_PATCH
+
+#if APPLESEED_VERSION_ALPHA != 0
```
We should always have `APPLESEED_VERSION_MATURITY` so this test will go away.

dictoon commented:
I like this a lot. There are a few issues to fix before it can be merged, but it's indeed much simpler and cleaner.

est77 commented:
I improved the PR. I did the following:

- Added a release type define, currently alpha.
- Generate the version header in foundation/core/version.h. foundation/core/appleseed is used in some places inside foundation, so it's better to leave it there.
- Added a renderer/api/version.h header that can be used by API clients.
- bumped the version number. we are now at 1.1.1 alpha.

dictoon commented on CMakeLists.txt:
```diff
@@ -122,30 +122,16 @@ endif ()
+set (appleseed_version_major        1)
+set (appleseed_version_minor        1)
+set (appleseed_version_patch        1)
+set (appleseed_version_release_type alpha)
```
I would really prefer `appleseed_version_maturity` since "release type" is very ambiguous.

We might also want to go directly to 1.2.0 (or higher) for the first stable release; issuing an initial release with a non-zero patch number seems weird to me :)

est77 replied:
I replaced release type by maturity.
About the version number, it's something we can decide later as the next release will be the first one to have this header.
